server/app/services/applicant/question/CurrencyQuestion.java

```diff
@@ -37,12 +38,18 @@ public ImmutableList<Path> getAllPaths() {
 
   @Override
   protected ImmutableMap<Path, ImmutableSet<ValidationErrorMessage>> getValidationErrorsInternal() {
-    // TODO(#1944): Validate that the provided currency is a valid number. Presently,
-    // this is only implemented in client-side validation.
+    // When staging updates, the attempt to update ApplicantData would have failed to
+    // convert to a currency and been noted as a failed update. We check for that here.
+    if (applicantQuestion.getApplicantData().getFailedUpdates().containsKey(getCurrencyPath())) {
```

Comment:
Passing this through feels a bit clunky to me. However, I believe we'd like to consolidate validation-related code to these classes. Presently, these are constructed after the controller has "staged" updates to the relevant ApplicantData.

It may be worth pursuing a separate type of refactoring where the candidate raw updates are passed here rather than staging them from the controller. I haven't thought that through enough yet to have confidence one way or another.

Assuming the current approach isn't too distasteful, I'd propose handling a refactoring such as that separately.

Reply:
That sounds good to me, I agree it's tricky to say which approach will be cleaner at moment.